pkg/lib/cockpit.d.ts

```diff
@@ -93,6 +93,49 @@ declare module 'cockpit' {
     function channel(options: ChannelOptions & { binary?: false; }): Channel<string>;
     function channel(options: ChannelOptions & { binary: true; }): Channel<Uint8Array>;
 
+    /* === cockpit.spawn ============================= */
+
+    interface DeferredPromise<T> extends Promise<T> {
+        /* jQuery Promise compatibility */
+        done(callback: (data: T) => void): DeferredPromise<T>
+        fail(callback: (exc: Error) => void): DeferredPromise<T>
+        always(callback: () => void): DeferredPromise<T>
+        progress(callback: (message: T, cancel: () => void) => void): DeferredPromise<T>
+    }
+
+    interface Deferred<T> {
+        resolve(): Deferred<T>;
+        reject(): Deferred<T>;
+        notify(): Deferred<T>;
+        promise: DeferredPromise<T>
+    }
+
+    interface Spawn<T> extends DeferredPromise<T> {
+        input(message: T, stream?: boolean): DeferredPromise<T>;
+        stream(callback: (data: T) => void): DeferredPromise<T>;
+        close(): void;
+    }
+
+    interface SpawnOptions {
+        binary?: boolean,
+        directory?: string;
+        err?: "out" | "ignore" | "message" | "pty";
+        environ?: string[];
+        pty?: boolean;
+        batch?: number;
+        latency?: number
+        superuser?: string;
```

Comment:
This could use the literal treatment: `"try" | "require"`.

I wonder if we should make some sort of a "common options" type for stuff like superuser, host, etc, and extend the specific channel options from that...

Reply:
Same, but avoided it until the other PR was merged :)